# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn import preprocessing

# Load in Data

In [2]:
credit_df = pd.read_csv("credit_card_churn.csv")

In [3]:
credit_df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

# Understanding the Data

In [4]:
credit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  object 
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                           

In [5]:
credit_df.info

<bound method DataFrame.info of        CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0      768805383  Existing Customer            45      M                3   
1      818770008  Existing Customer            49      F                5   
2      713982108  Existing Customer            51      M                3   
3      769911858  Existing Customer            40      F                4   
4      709106358  Existing Customer            40      M                3   
...          ...                ...           ...    ...              ...   
10122  772366833  Existing Customer            50      M                2   
10123  710638233  Attrited Customer            41      M                2   
10124  716506083  Attrited Customer            44      F                1   
10125  717406983  Attrited Customer            30      M                2   
10126  714337233  Attrited Customer            43      F                2   

      Education_Level Marital_Status Income

In [6]:
credit_df.describe()

CLIENTNUM  Customer_Age  Dependent_count  Months_on_book  \
count  1.012700e+04  10127.000000     10127.000000    10127.000000   
mean   7.391776e+08     46.325960         2.346203       35.928409   
std    3.690378e+07      8.016814         1.298908        7.986416   
min    7.080821e+08     26.000000         0.000000       13.000000   
25%    7.130368e+08     41.000000         1.000000       31.000000   
50%    7.179264e+08     46.000000         2.000000       36.000000   
75%    7.731435e+08     52.000000         3.000000       40.000000   
max    8.283431e+08     73.000000         5.000000       56.000000   

       Total_Relationship_Count  Months_Inactive_12_mon  \
count              10127.000000            10127.000000   
mean                   3.812580                2.341167   
std                    1.554408                1.010622   
min                    1.000000                0.000000   
25%                    3.000000                2.000000   
50%                    4.000000                2.000000   
75%                    5.000000                3.000000   
max                    6.000000                6.000000   

       Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
count           10127.000000  10127.000000         10127.000000   
mean                2.455317   8631.953698          1162.814061   
std                 1.106225   9088.776650           814.987335   
min                 0.000000   1438.300000             0.000000   
25%                 2.000000   2555.000000           359.000000   
50%                 2.000000   4549.000000          1276.000000   
75%                 3.000000  11067.500000          1784.000000   
max                 6.000000  34516.000000          2517.000000   

       Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
count     10127.000000          10127.000000     10127.000000    10127.000000   
mean       7469.139637              0.759941      4404.086304       64.858695   
std        9090.685324              0.219207      3397.129254       23.472570   
min           3.000000              0.000000       510.000000       10.000000   
25%        1324.500000              0.631000      2155.500000       45.000000   
50%        3474.000000              0.736000      3899.000000       67.000000   
75%        9859.000000              0.859000      4741.000000       81.000000   
max       34516.000000              3.397000     18484.000000      139.000000   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
count         10127.000000           10127.000000   
mean              0.712222               0.274894   
std               0.238086               0.275691   
min               0.000000               0.000000   
25%               0.582000               0.023000   
50%               0.702000               0.176000   
75%               0.818000               0.503000   
max               3.714000               0.999000   

       Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
count                                       10127.000000                                                                                    
mean                                            0.159997                                                                                    
std                                             0.365301                                                                                    
min                                             0.000008                                                                                    
25%                                             0.000099                                                                                    
50%                                             0.000181                                                                                    
75%                                             

# Data Cleaning

In [7]:
#Drop irrelevant columns 
credit_df = credit_df.drop(columns=['Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
               'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'])

In [8]:
credit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Attrition_Flag            10127 non-null  object 
 2   Customer_Age              10127 non-null  int64  
 3   Gender                    10127 non-null  object 
 4   Dependent_count           10127 non-null  int64  
 5   Education_Level           10127 non-null  object 
 6   Marital_Status            10127 non-null  object 
 7   Income_Category           10127 non-null  object 
 8   Card_Category             10127 non-null  object 
 9   Months_on_book            10127 non-null  int64  
 10  Total_Relationship_Count  10127 non-null  int64  
 11  Months_Inactive_12_mon    10127 non-null  int64  
 12  Contacts_Count_12_mon     10127 non-null  int64  
 13  Credit_Limit              10127 non-null  float64
 14  Total_

There are 21 columns with 10127 observations in total. No columns have any nulls present, so no cleaning is required on that end. There are multiple variables within this data set that are not numbers and need to be encoded as such so that they can be fed into our developed models and run properly. There are six different object variables.

Object variables that need to be cleaned / one-hot encoded:
- Attrition Flag
- Gender
- Education Level
- Marital Status 
- Income Category 
- Card Category 

In [9]:
credit_df['Attrition_Flag'].value_counts()

Existing Customer    8500
Attrited Customer    1627
Name: Attrition_Flag, dtype: int64

In [10]:
credit_df['Attrition_Flag'] = credit_df['Attrition_Flag'].map({'Existing Customer': 1, 'Attrited Customer': 0})

In [11]:
credit_df['Education_Level'].value_counts()

Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: Education_Level, dtype: int64

In [12]:
credit_df['Marital_Status'].value_counts()

Married     4687
Single      3943
Unknown      749
Divorced     748
Name: Marital_Status, dtype: int64

In [13]:
credit_df['Income_Category'].value_counts()

Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: Income_Category, dtype: int64

In [14]:
credit_df['Card_Category'].value_counts()

Blue        9436
Silver       555
Gold         116
Platinum      20
Name: Card_Category, dtype: int64

In [15]:
credit_df['Gender'].value_counts()

F    5358
M    4769
Name: Gender, dtype: int64

In [16]:
credit_df_cleaned = pd.get_dummies(credit_df, columns = ['Marital_Status', 'Gender'], dtype=int)

In [17]:
credit_df_cleaned

CLIENTNUM  Attrition_Flag  Customer_Age  Dependent_count  \
0      768805383               1            45                3   
1      818770008               1            49                5   
2      713982108               1            51                3   
3      769911858               1            40                4   
4      709106358               1            40                3   
...          ...             ...           ...              ...   
10122  772366833               1            50                2   
10123  710638233               0            41                2   
10124  716506083               0            44                1   
10125  717406983               0            30                2   
10126  714337233               0            43                2   

      Education_Level Income_Category Card_Category  Months_on_book  \
0         High School     $60K - $80K          Blue              39   
1            Graduate  Less than $40K          Blue              44   
2            Graduate    $80K - $120K          Blue              36   
3         High School  Less than $40K          Blue              34   
4          Uneducated     $60K - $80K          Blue              21   
...               ...             ...           ...             ...   
10122        Graduate     $40K - $60K          Blue              40   
10123         Unknown     $40K - $60K          Blue              25   
10124     High School  Less than $40K          Blue              36   
10125        Graduate     $40K - $60K          Blue              36   
10126        Graduate  Less than $40K        Silver              25   

       Total_Relationship_Count  Months_Inactive_12_mon  ...  Total_Trans_Amt  \
0                             5                       1  ...             1144   
1                             6                       1  ...             1291   
2                             4                       1  ...             1887   
3                             3                       4  ...             1171   
4                             5                       1  ...              816   
...                         ...                     ...  ...              ...   
10122                         3                       2  ...            15476   
10123                         4                       2  ...             8764   
10124                         5                       3  ...            10291   
10125                         4                       3  ...             8395   
10126                         6                       2  ...            10294   

       Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
0                  42                1.625                  0.061   
1                  33                3.714                  0.105   
2                  20                2.333                  0.000   
3                  20                2.333                  0.760   
4                  28                2.500                  0.000   
...               ...                  ...                    ...   
10122             117                0.857                  0.462   
10123              69                0.683                  0.511   
10124              60                0.818                  0.000   
10125              62                0.722                  0.000   
10126              61                0.649                  0.189   

       Marital_Status_Divorced  Marital_Status_Married  Marital_Status_Single  \
0                            0                       1                      0   
1                            0                       0                      1   
2                            0                       1                      0   
3                            0                       0                      0   
4                            0                       1                      0   
...                        ...                     ...       

In [18]:
credit_df_test = credit_df_cleaned.copy()

In [19]:

  
# label_encoder object knows 
# how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
  
# Encode labels in column 'species'.
credit_df_test['Income_Category_Labels']= label_encoder.fit_transform(credit_df_test['Income_Category'])
  
credit_df_test['Income_Category_Labels'].unique()

array([2, 4, 3, 1, 0, 5])

In [20]:
credit_df_test[['Income_Category', 'Income_Category_Labels']].head(25)

Income_Category  Income_Category_Labels
0      $60K - $80K                       2
1   Less than $40K                       4
2     $80K - $120K                       3
3   Less than $40K                       4
4      $60K - $80K                       2
5      $40K - $60K                       1
6          $120K +                       0
7      $60K - $80K                       2
8      $60K - $80K                       2
9     $80K - $120K                       3
10         $120K +                       0
11     $40K - $60K                       1
12    $80K - $120K                       3
13     $60K - $80K                       2
14  Less than $40K                       4
15    $80K - $120K                       3
16    $80K - $120K                       3
17    $80K - $120K                       3
18     $40K - $60K                       1
19         Unknown                       5
20     $60K - $80K                       2
21  Less than $40K                       4
22     $40K - $60K                       1
23  Less than $40K                       4
24    $80K - $120K                       3

Marital status and gender are object variables with no clear order that we can interpret numerically, so for these two it makes the most sense to one-hot encode and create dummy variables for each value. However, income category, card category, and education level each have a natural ranking and should be label encoded. The label encoder function, by default, ranks the object variable based on alphabetical value - this would be a potential cause for confusion when it comes time to interpret. To avoid this mislabeling we are manually encoding these labels.

In [21]:
credit_df_test['Card_Category'] = credit_df_test['Card_Category'].apply(lambda x: ['Blue', 'Silver', 'Gold', 'Platinum'].index(x))
credit_df_test['Card_Category'].value_counts()

0    9436
1     555
2     116
3      20
Name: Card_Category, dtype: int64

In the variables, education level and income category, there are 'Unknown' values - these won't provide much benefit to our models. If there are observations with multiple 'Unknown' values for multiple categories then it may be a good idea to drop these observations. Otherwise, 'Unknown' will be coded as the numerical value of 0. 

In [22]:
credit_df_test['Education_Level'].value_counts()

Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: Education_Level, dtype: int64

In [23]:
credit_df_test['Education_Level'] = credit_df_test['Education_Level'].apply(lambda x: ['Unknown', 'Uneducated', 'High School', 'College', 'Graduate', 'Post-Graduate', 'Doctorate'].index(x))
credit_df_test['Education_Level'].value_counts()

4    3128
2    2013
0    1519
1    1487
3    1013
5     516
6     451
Name: Education_Level, dtype: int64

In [24]:
credit_df_test['Income_Category'].value_counts()

Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: Income_Category, dtype: int64

In [25]:
credit_df_test['Income_Category'] = credit_df_test['Income_Category'].apply(lambda x: ['Unknown', 'Less than $40K', '$40K - $60K', '$60K - $80K', '$80K - $120K', '$120K +'].index(x))
credit_df_test['Income_Category'].value_counts()

1    3561
2    1790
4    1535
3    1402
0    1112
5     727
Name: Income_Category, dtype: int64

Looking into the 'Unknown' values to see if there are any observations with multiple pieces of unknown data. 

In [26]:
credit_df

CLIENTNUM  Attrition_Flag  Customer_Age Gender  Dependent_count  \
0      768805383               1            45      M                3   
1      818770008               1            49      F                5   
2      713982108               1            51      M                3   
3      769911858               1            40      F                4   
4      709106358               1            40      M                3   
...          ...             ...           ...    ...              ...   
10122  772366833               1            50      M                2   
10123  710638233               0            41      M                2   
10124  716506083               0            44      F                1   
10125  717406983               0            30      M                2   
10126  714337233               0            43      F                2   

      Education_Level Marital_Status Income_Category Card_Category  \
0         High School        Married     $60K - $80K          Blue   
1            Graduate         Single  Less than $40K          Blue   
2            Graduate        Married    $80K - $120K          Blue   
3         High School        Unknown  Less than $40K          Blue   
4          Uneducated        Married     $60K - $80K          Blue   
...               ...            ...             ...           ...   
10122        Graduate         Single     $40K - $60K          Blue   
10123         Unknown       Divorced     $40K - $60K          Blue   
10124     High School        Married  Less than $40K          Blue   
10125        Graduate        Unknown     $40K - $60K          Blue   
10126        Graduate        Married  Less than $40K        Silver   

       Months_on_book  ...  Months_Inactive_12_mon  Contacts_Count_12_mon  \
0                  39  ...                       1                      3   
1                  44  ...                       1                      2   
2                  36  ...                       1                      0   
3                  34  ...                       4                      1   
4                  21  ...                       1                      0   
...               ...  ...                     ...                    ...   
10122              40  ...                       2                      3   
10123              25  ...                       2                      3   
10124              36  ...                       3                      4   
10125              36  ...                       3                      3   
10126              25  ...                       2                      4   

       Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0           12691.0                  777          11914.0   
1            8256.0                  864           7392.0   
2            3418.0                    0           3418.0   
3            3313.0                 2517            796.0   
4            4716.0                    0           4716.0   
...             ...                  ...              ...   
10122        4003.0                 1851           2152.0   
10123        4277.0                 2186           2091.0   
10124        5409.0                    0           5409.0   
10125        5281.0                    0           5281.0   
10126       10388.0                 1961           8427.0   

       Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
0                     1.335             1144              42   
1                     1.541             1291              33   
2                     2.594             1887              20   
3                     1.405             1171              20   
4                     2.175              816              28   
...                     ...              ...             ...   
10122                 0.703            15476             117   
10123                 0.804             8764              69   
10124                 0.819            1029

In [27]:
unknown_3 = credit_df[['Marital_Status', 'Education_Level', 'Income_Category']]
unknown_3['matching'] = unknown_3.apply(lambda x: x.Marital_Status == x.Education_Level == x.Income_Category, axis=1)


<ipython-input-27-ae8f0e07844b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  unknown_3['matching'] = unknown_3.apply(lambda x: x.Marital_Status == x.Education_Level == x.Income_Category, axis=1)


In [28]:
unknown_3['matching'].value_counts()

False    10120
True         7
Name: matching, dtype: int64

In [29]:
unknown_3.loc[unknown_3['matching'] == 'True']

Empty DataFrame
Columns: [Marital_Status, Education_Level, Income_Category, matching]
Index: []

There are only 7 cases where marital status, education level, and income category have the same value. Even if all 7 of these cases were matching on 'Unknown', it is still a small enough number where we don't have to worry about them significantly skewing our data and impacting our model performance. We will continue with the full dataset and not drop any observations. 

# Exploratory Data Analysis

# Model Building

The purpose of the model building stage is to create a highly function classification model that can take in the numerous pieces of customer data and classify whether the customer is likely to churn / cancel their credit card. There are multiple routes that could generate a production grade model that could signficantly help the bank to understand what makes a customer likely to leave - in turn this could impact future marketing strategies, as well as allow them to be proactive and attempt to find methods of mitigating this issue among certain groups. 

## Logistic Regression

## Decision Tree

## Random Forest

## RF XGBoost

## RF AdaBoost

## Convolutional Neural Network (CNN)